packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -264,6 +264,23 @@ function attachShadow(vm: VM) {
     }
 }
 
+function warnForApiCalledInConstructor(methodName: string) {
+    return function (this: LightningElement, ...args: any[]) {
+        const vm = getAssociatedVM(this);
+        const { elm, renderer } = vm;
+
+        if (process.env.NODE_ENV !== 'production' && isBeingConstructed(vm)) {
+            logError(
+                `this.${methodName}() should not be called during the construction of the custom element for ${getComponentTag(
+                    vm
+                )} because the element is not yet in the DOM or has no children yet.`
+            );
+        }
+
+        return (renderer as any)[methodName](elm, ...args);
```

Comment:
I like the overall direction of the PR and also the fact that it reduces the overall code size. The main issue that I have is that the PR doesn't preserve type safety and makes it harder to understand which API the `LightningElement` implements.

What do you think about keeping all the methods defined inlined in the `LightningElement.prototype` object and remove the duplicate code by moving the warning into a separate function?

```ts
LightningElement.prototype = {
    // [...]
 
    getBoundingClientRect(): ClientRect {
        const vm = getAssociatedVM(this);
        const {
            elm,
            renderer: { getBoundingClientRect },
        } = vm;

        if (process.env.NODE_ENV !== 'production') {
            this.warnIfInvokedDuringConstruction();
        }

        return getBoundingClientRect(elm);
    },

    // [...]
}
```

Reply:
@pmdartus This doesn't reduce the duplication nearly as much as my approach. When UX conflicts with DX, I tend to favor UX (in this case, reduced code size vs TypeScript type safety).

I wonder if there is some way to annotate the types with JSDoc comments or something. I'll look into this.

Reply:
How much are we saving? While I can relate with your sentiment, I think there's something to be said for code quality and its contribution to UX as well. My initial impression is that keeping type-safety is worth the tradeoff.

Reply:
That's fair. Looks like it saves 227 bytes minified and 2 bytes min+gz (gzip is good at deduplicating 😛). So you're right, I shouldn't oversell the savings here.